src/java.base/share/classes/sun/security/util/DerValue.java

```diff
@@ -227,106 +249,166 @@ public DerValue(String value) {
+    private static byte[] string2bytes(byte stringTag, String value) {
+        Charset charset = switch (stringTag) {
+            case tag_PrintableString, tag_IA5String, tag_GeneralString -> US_ASCII;
+            case tag_T61String -> ISO_8859_1;
+            case tag_BMPString -> UTF_16BE;
+            case tag_UTF8String -> UTF_8;
+            case tag_UniversalString -> Charset.forName("UTF_32BE");
+            default -> throw new IllegalArgumentException("Unsupported DER string type");
+        };
+        return value.getBytes(charset);
+    }
+
+    DerValue(byte tag, byte[] buffer, boolean allowBER) {
+        this(tag, buffer, 0, buffer.length, allowBER);
     }
 
     /**
      * Creates a DerValue from a tag and some DER-encoded data.
      *
+     * This is a public constructor.
+     *
      * @param tag the DER type tag
-     * @param data the DER-encoded data
+     * @param buffer the DER-encoded data
      */
-    public DerValue(byte tag, byte[] data) {
-        this(tag, data, true);
-    }
-
-    /*
-     * package private
-     */
-    DerValue(DerInputBuffer in) throws IOException {
-
-        // XXX must also parse BER-encoded constructed
-        // values such as sequences, sets...
-        tag = (byte)in.read();
-        byte lenByte = (byte)in.read();
-        length = DerInputStream.getLength(lenByte, in);
-        if (length == -1) {  // indefinite length encoding found
-            DerInputBuffer inbuf = in.dup();
-            inbuf = new DerInputBuffer(
-                    DerIndefLenConverter.convertStream(inbuf, lenByte, tag),
-                    in.allowBER);
-            if (tag != inbuf.read())
-                throw new IOException
-                        ("Indefinite length encoding not supported");
-            length = DerInputStream.getDefiniteLength(inbuf);
-            buffer = inbuf.dup();
-            buffer.truncate(length);
-            data = new DerInputStream(buffer);
-            // indefinite form is encoded by sending a length field with a
-            // length of 0. - i.e. [1000|0000].
-            // the object is ended by sending two zero bytes.
-            in.skip(length + 2);
-        } else {
-
-            buffer = in.dup();
-            buffer.truncate(length);
-            data = new DerInputStream(buffer);
-
-            in.skip(length);
-        }
-    }
-
-    // Get an ASN.1/DER encoded datum from a buffer w/ additional
-    // arg to control whether DER checks are enforced.
-    DerValue(byte[] buf, boolean allowBER) throws IOException {
-        data = init(true, new ByteArrayInputStream(buf), allowBER);
+    public DerValue(byte tag, byte[] buffer) {
+        this(tag, buffer.clone(), true);
     }
 
     /**
-     * Get an ASN.1/DER encoded datum from a buffer.  The
-     * entire buffer must hold exactly one datum, including
-     * its tag and length.
+     * Parse an ASN.1/BER encoded datum. The entire encoding must hold exactly
```

Comment:
Yes. In fact I'm not quite confident on our usage of allowBER. In a lot of cases we are actually quite strict. Since this code change is meant to be a refactoring and does not intend to fix many things, I don't intent to make many behavior change.

Reply:
We have to be strict in "sensitive" area such as signatures. The parsing code seems to be still mainly DER. It was never fully BER, but just some. It's good to keep behavior change minimum as this is like a re-write and may already have some unintentional changes.

Reply:
As this constructor calls another with allowBER=true, I leave the BER word there. Yes, I understand our impl is not always precise. To make it so, we need to adjust the callers everywhere to match the RFCs, but then there will be compatibility issues.

Reply:
Yes, I agree.